README.md
# holiday_jp-dart

[![Pub Version](https://img.shields.io/pub/v/holiday_jp?color=blueviolet)](https://pub.dev/packages/holiday_jp)
[![test](https://github.com/wapa5pow/holiday_jp-dart/actions/workflows/ci.yml/badge.svg)](https://github.com/wapa5pow/holiday_jp-dart/actions/workflows/ci.yml)
[![latest_holiday_jp](https://github.com/wapa5pow/holiday_jp-dart/actions/workflows/latest_holiday_jp.yml/badge.svg)](https://github.com/wapa5pow/holiday_jp-dart/actions/workflows/latest_holiday_jp.yml)

🎌 Japanese holiday for Dart

This library uses [holiday_jp](https://github.com/holiday-jp/holiday_jp) as Japanese holiday dataset.

## Example

```dart
import 'package:holiday_jp/holiday_jp.dart' as holiday_jp;
import 'package:intl/intl.dart';

final _formatter = DateFormat('yyyy-MM-dd');

void main(List<String> arguments) {
  final marineDay = DateTime.utc(2020, 7, 23);

  final holiday = holiday_jp.getHoliday(marineDay);
  if (holiday == null) {
    throw Exception('holiday mus not be null');
  }
  print('${holiday.date} is ${holiday.nameEn} in Japan.\n');

  final isMarineDayHoliday = holiday_jp.isHoliday(marineDay);
  print(
      'The question whether Marine Day is holiday or not is $isMarineDayHoliday.\n');

  final start = DateTime.utc(2022, 1, 1);
  final end = DateTime.utc(2022, 1, 15);
  final holidays = holiday_jp.between(start, end);
  print(
      'Holidays between ${_formatter.format(start)} and ${_formatter.format(end)} are the followings.');
  for (final holiday in holidays) {
    print('- ${holiday.date} is ${holiday.nameEn}');
  }
}
```

## Tools

To update generated holiday file, run the following.

```bash
./tool/update_holidays.sh
```

CHANGELOG.md
## 0.0.8

- Remove intl package

## 0.0.7

- Use lower intl package version.

## 0.0.6

- Make example public.

## 0.0.5

- Fix readme.

## 0.0.4

- Fix package description.

## 0.0.3

- Fix package layout.

## 0.0.2

- Fix library publicity.

## 0.0.1

- Initial version.


test/holiday_test.dart
import 'package:holiday_jp/src/holiday.dart';
import 'package:test/test.dart';

void main() {
  final marineDay = DateTime.utc(2020, 7, 23);

  group('$Holiday', () {
    final holiday1 = Holiday(
      date: '2022-07-23',
      week: '木',
      weekEn: 'Thursday',
      name: '海の日',
      nameEn: 'Marine Day',
    );
    final holiday2 = Holiday(
      date: '2022-07-23',
      week: '木',
      weekEn: 'Thursday',
      name: '海の日',
      nameEn: 'Marine Day',
    );
    final holiday3 = Holiday(
      date: '1970-01-01',
      week: '木',
      weekEn: 'Thursday',
      name: '元旦',
      nameEn: 'New Year\'s Day',
    );

    test('equality', () {
      expect(holiday1, holiday2);
      expect(holiday1, isNot(holiday3));
    });
  });

  group('getHoliday', () {
    test('Marine Day is holiday', () {
      final holiday = getHoliday(marineDay);
      expect(holiday, isNotNull);
      expect(holiday!.date, '2020-07-23');
      expect(holiday.week, '木');
      expect(holiday.weekEn, 'Thursday');
      expect(holiday.name, '海の日');
      expect(holiday.nameEn, 'Marine Day');
    });

    test('1970-01-02 is not holiday', () {
      expect(getHoliday(DateTime.utc(1970, 1, 2)), isNull);
    });
  });

  group('isHoliday', () {
    final testCases = [
      [DateTime.utc(1970, 1, 1), true],
      [DateTime.utc(1970, 1, 1, 1, 1, 1, 1), true],
      [DateTime.utc(1970, 1, 2, 1, 1, 1, 1), false],
      [DateTime.utc(1970, 1, 2, 0, 0, 0, 0), false],
      [marineDay, true],
    ];

    for (final testCase in testCases) {
      final input = testCase[0] as DateTime;
      final want = testCase[1] as bool;
      test('Check $input is holiday or not', () {
        expect(isHoliday(input), want);
      });
    }
  });

  group('between', () {
    final testCases = [
      [
        DateTime.utc(1970, 1, 1, 1, 1, 1, 1),
        DateTime.utc(1970, 1, 1, 1, 1, 1, 1),
        [
          Holiday(
            date: '1970-01-01',
            week: '木',
            weekEn: 'Thursday',
            name: '元日',
            nameEn: 'New Year\'s Day',
          ),
        ]
      ],
      [
        DateTime.utc(1970, 1, 1, 1, 1, 1, 1),
        DateTime.utc(1970, 1, 2, 1, 1, 1, 1),
        [
          Holiday(
            date: '1970-01-01',
            week: '木',
            weekEn: 'Thursday',
            name: '元日',
            nameEn: 'New Year\'s Day',
          ),
        ]
      ],
      [
        DateTime.utc(1970, 1, 1, 1, 1, 1, 1),
        DateTime.utc(1970, 1, 14, 1, 1, 1, 1),
        [
          Holiday(
            date: '1970-01-01',
            week: '木',
            weekEn: 'Thursday',
            name: '元日',
            nameEn: 'New Year\'s Day',
          ),
        ]
      ],
      [
        DateTime.utc(1970, 1, 1, 1, 1, 1, 1),
        DateTime.utc(1970, 1, 15, 1, 1, 1, 1),
        [
          Holiday(
            date: '1970-01-01',
            week: '木',
            weekEn: 'Thursday',
            name: '元日',
            nameEn: 'New Year\'s Day',
          ),
          Holiday(
            date: '1970-01-15',
            week: '木',
            weekEn: 'Thursday',
            name: '成人の日',
            nameEn: 'Coming of Age Day',
          ),
        ]
      ],
      [
        marineDay,
        marineDay,
        [
          Holiday(
            date: '2020-07-23',
            week: '木',
            weekEn: 'Thursday',
            name: '海の日',
            nameEn: 'Marine Day',
          ),
        ]
      ],
    ];

    for (final testCase in testCases) {
      final start = testCase[0] as DateTime;
      final end = testCase[1] as DateTime;
      final want = testCase[2] as List<Holiday>;
      test('from $start to $end', () {
        expect(between(start, end), want);
      });
    }
  });
}


test/holiday_jp_dart_test.dart
import 'package:holiday_jp_dart/holiday_jp_dart.dart';
import 'package:test/test.dart';

void main() {
  test('calculate', () {
    expect(calculate(), 42);
  });
}


tool/generate.dart
import 'dart:convert';
import 'dart:io';

// ignore: depend_on_referenced_packages
import 'package:yaml/yaml.dart';

/// Generate holiday.g.dart using holiday_jp dataset
void main() {
  final document =
      loadYaml(File('holiday_jp/holidays_detailed.yml').readAsStringSync())
          as YamlMap;
  final dates = (document.keys.toList())..sort();

  // generate holidays
  final sink = File('lib/src/holiday.g.dart').openWrite();
  sink.add(utf8.encode("""
part of 'holiday.dart';

final holidays = {
"""));

  for (final date in dates) {
    final holiday = document[date] as YamlMap;
    sink.add(utf8.encode("""
  "${holiday["date"]}": Holiday(
    date: "${holiday['date']}",
    week: "${holiday['week']}",
    weekEn: "${holiday['week_en']}",
    name: "${holiday['name']}",
    nameEn: "${holiday['name_en']}",
  ),
"""));
  }

  sink.add(utf8.encode("};\n"));
}


lib/src/holiday.dart
part 'holiday.g.dart';

class Holiday {
  final String date;
  final String week;
  final String weekEn;
  final String name;
  final String nameEn;

  Holiday({
    required this.date,
    required this.week,
    required this.weekEn,
    required this.name,
    required this.nameEn,
  });

  @override
  bool operator ==(Object other) =>
      other is Holiday &&
      other.date == date &&
      other.week == week &&
      other.weekEn == weekEn &&
      other.name == name &&
      other.nameEn == nameEn;

  @override
  int get hashCode => Object.hash(
        date,
        week,
        weekEn,
        name,
        nameEn,
      );
}

/// Return [Holiday]
///
/// If [dateTime] is not holiday, it returns null.
Holiday? getHoliday(DateTime dateTime) {
  final date =
      '${dateTime.year}-${dateTime.month.toString().padLeft(2, '0')}-${dateTime.day.toString().padLeft(2, '0')}';
  return holidays.containsKey(date) ? holidays[date] : null;
}

/// Return true if [dateTime] is holiday
bool isHoliday(DateTime dateTime) {
  return getHoliday(dateTime) != null;
}

/// Find [Holiday]s between parameters
List<Holiday> between(DateTime start, DateTime end) {
  final results = <Holiday>[];
  for (;; start = start.add(Duration(days: 1))) {
    if (start.isAfter(end)) {
      break;
    }
    final holiday = getHoliday(start);
    if (holiday != null) {
      results.add(holiday);
    }
  }
  return results;
}


example/main.dart
import 'package:holiday_jp/holiday_jp.dart' as holiday_jp;
import 'package:intl/intl.dart';

final _formatter = DateFormat('yyyy-MM-dd');

void main(List<String> arguments) {
  final marineDay = DateTime.utc(2020, 7, 23);

  final holiday = holiday_jp.getHoliday(marineDay);
  if (holiday == null) {
    throw Exception('holiday mus not be null');
  }
  print('${holiday.date} is ${holiday.nameEn} in Japan.\n');

  final isMarineDayHoliday = holiday_jp.isHoliday(marineDay);
  print(
      'The question whether Marine Day is holiday or not is $isMarineDayHoliday.\n');

  final start = DateTime.utc(2022, 1, 1);
  final end = DateTime.utc(2022, 1, 15);
  final holidays = holiday_jp.between(start, end);
  print(
      'Holidays between ${_formatter.format(start)} and ${_formatter.format(end)} are the followings.');
  for (final holiday in holidays) {
    print('- ${holiday.date} is ${holiday.nameEn}');
  }
}
